Sources/GRPCNIOTransportCore/Internal/PercentEncoding.swift

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright 2024, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package enum PercentEncoding {
+  package static func encodeAuthority(_ input: String) -> String {
+    Self.encode(input) {
+      Self.isAuthorityChar($0)
+    }
+  }
+
+  package static func encode(
+    _ input: String,
+    isValidCharacter: (UInt8) -> Bool
+  ) -> String {
+    var output: [UInt8] = []
+    output.reserveCapacity(input.utf8.count)
+
+    for char in input.utf8 {
+      if isValidCharacter(char) {
+        output.append(char)
+      } else {
+        output.append(UInt8(ascii: "%"))
```

Comment:
Will the `UInt8` conversion of `%` be computed and stored at compile time? Is it worth doing that ourselves and hardcoding it e.g. `UInt8(0x25) // %`